File: Java/Medium/Maximum Matching of Players With Trainers.java
class Solution {
    public int matchPlayersAndTrainers(int[] players, int[] trainers) {
        Arrays.sort(players);
        Arrays.sort(trainers);

        int P = players.length, T = trainers.length;

        int count = 0;
        int j = 0;

        // For each player, find the least capable trainer that can train him (greedy)
        for(int i = 0; i < P; i++) {
            while(j < T && trainers[j] < players[i])
                j++;

            if (j >= T)
                break;  // No suitable trainer found

            count++;
            j++;  // Move to next trainer to avoid reusing him
        }

        return count;
    }
}

File: Python/Medium/Maximum Matching of Players With Trainers.py
class Solution:
    def matchPlayersAndTrainers(self, players: List[int], trainers: List[int]) -> int:
        players.sort()
        trainers.sort()

        count = 0
        j = 0  # pointer for trainer array
        # For each player, find the least capable trainer that can train him (greedy)
        for i in range(len(players)):
            while j < len(trainers) and trainers[j] < players[i]:
                j += 1

            if j >= len(trainers):
                break  # No suitable trainer found

            
            count += 1
            j += 1  # Move to next trainer to avoid reusing him

        return count
